aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsHttpServletRequestHelper.java

```diff
@@ -0,0 +1,452 @@
+        asyncContext = new AwsAsyncContext((HttpServletRequest) servletRequest, (HttpServletResponse) servletResponse, containerHandler);
+        log.debug("Starting async context for request: " + SecurityUtils.crlf(requestId));
+        return asyncContext;
+    }
+
+    public static AsyncContext getAsyncContext(AwsAsyncContext asyncContext, String requestId) {
+        if (asyncContext == null) {
+            throw new IllegalStateException("Request " + SecurityUtils.crlf(requestId)
+                    + " is not in asynchronous mode. Call startAsync before attempting to get the async context.");
+        }
+        return asyncContext;
+    }
+
+
+
+
+    protected static String getFirst(Map<String, List<String>> map, String key) {
+        List<String> values = map.get(key);
+        if (values == null || values.size() == 0) {
+            return null;
+        }
+        return values.get(0);
+    }
+
+    protected void putSingle(Map<String, List<String>> map, String key, String value) {
+        List<String> values = findKey(map, key);
+        values.clear();
+        values.add(value);
+    }
+
+    protected List<String> findKey(Map<String, List<String>> map, String key) {
+        List<String> values = map.get(key);
+        if (values == null) {
+            values = new ArrayList<>();
+            map.put(key, values);
+        }
+        return values;
+    }
+
+    /**
+     * Given the Cookie header value, parses it and creates a Cookie object
+     * @param headerValue The string value of the HTTP Cookie header
+     * @return An array of Cookie objects from the header
+     */
+    protected static Cookie[] parseCookieHeaderValue(String headerValue) {
+        List<AwsHttpServletRequest.HeaderValue> parsedHeaders = parseHeaderValue(headerValue,  ";", ",");
+
+        return parsedHeaders.stream()
+                .filter(e -> e.getKey() != null)
+                .map(e -> new Cookie(SecurityUtils.crlf(e.getKey()), SecurityUtils.crlf(e.getValue())))
+                .toArray(Cookie[]::new);
+    }
+
+    /**
+     * Generic method to parse an HTTP header value and split it into a list of key/values for all its components.
+     * When the property in the header does not specify a key the key field in the output pair is null and only the value
+     * is populated. For example, The header <code>Accept: application/json; application/xml</code> will contain two
+     * key value pairs with key null and the value set to application/json and application/xml respectively.
+     *
+     * @param headerValue The string value for the HTTP header
+     * @param valueSeparator The separator to be used for parsing header values
+     * @return A list of SimpleMapEntry objects with all of the possible values for the header.
+     */
+    protected static List<AwsHttpServletRequest.HeaderValue> parseHeaderValue(String headerValue, String valueSeparator, String qualifierSeparator) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 439-447 into a separate method.